src/index.ts

```diff
@@ -0,0 +1,449 @@
+      derivedKey.slice(0, 16),
+      Buffer.from(json.crypto.cipherparams.iv, 'hex'),
+    )
+    const seed = runCipherBuffer(decipher, ciphertext)
+    return new Wallet(seed)
+  }
+
+  /*
+   * Based on https://github.com/ethereum/pyethsaletool/blob/master/pyethsaletool.py
+   * JSON fields: encseed, ethaddr, btcaddr, email
+   */
+  public static fromEthSale(input: string | Object, password: string): Wallet {
+    const json = typeof input === 'object' ? input : JSON.parse(input)
+
+    const encseed = Buffer.from(json.encseed, 'hex')
+
+    // key derivation
+    const derivedKey = crypto.pbkdf2Sync(password, password, 2000, 32, 'sha256').slice(0, 16)
+
+    // seed decoding (IV is first 16 bytes)
+    // NOTE: crypto (derived from openssl) when used with aes-*-cbc will handle PKCS#7 padding internally
+    //       see also http://stackoverflow.com/a/31614770/4964819
+    const decipher = crypto.createDecipheriv('aes-128-cbc', derivedKey, encseed.slice(0, 16))
+    const seed = runCipherBuffer(decipher, encseed.slice(16))
+
+    const wallet = new Wallet(ethUtil.keccak256(seed))
+    if (wallet.getAddress().toString('hex') !== json.ethaddr) {
+      throw new Error('Decoded key mismatch - possibly wrong passphrase')
+    }
+    return wallet
+  }
+
+  // private getters
+
+  private get pubKey(): Buffer {
+    if (!keyExists(this.publicKey)) {
+      this.publicKey = ethUtil.privateToPublic(this.privateKey as Buffer)
+    }
+    return this.publicKey
+  }
+
+  private get privKey(): Buffer {
+    if (!keyExists(this.privateKey)) {
+      throw new Error('This is a public key only wallet')
+    }
+    return this.privateKey
+  }
+
+  constructor(
+    private readonly privateKey?: Buffer | undefined,
+    private publicKey: Buffer | undefined = undefined,
+  ) {
+    if (privateKey && publicKey) {
+      throw new Error('Cannot supply both a private and a public key to the constructor')
+    }
+
+    if (!privateKey && !publicKey) {
+    }
+
+    if (privateKey && !ethUtil.isValidPrivate(privateKey)) {
+      throw new Error('Private key does not satisfy the curve requirements (ie. it is invalid)')
+    }
+
+    if (publicKey && !ethUtil.isValidPublic(publicKey)) {
+      throw new Error('Invalid public key')
+    }
+  }
+
+  // public instance methods
+
+  public getPrivateKey(): Buffer {
+    return this.privKey
+  }
+
+  public getPrivateKeyString(): string {
+    return ethUtil.bufferToHex(this.privKey)
+  }
+
+  public getPublicKey(): Buffer {
+    return this.pubKey
+  }
+
+  public getPublicKeyString(): string {
+    return ethUtil.bufferToHex(this.getPublicKey())
+  }
+
+  public getAddress(): Buffer {
+    return ethUtil.publicToAddress(this.pubKey)
+  }
+
+  public getAddressString(): string {
+    return ethUtil.bufferToHex(this.getAddress())
+  }
+
+  public getChecksumAddressString(): string {
+    return ethUtil.toChecksumAddress(this.getAddressString())
+  }
+
+  public toV3(password: string, opts?: Partial<V3Params>) {
```

Comment:
I remember seeing a special syntax for default values along the line of: `{ n: 262144, r: 8, p: 1, ... }: V3Params = {}`. In this case however the number of attributes is high, so it would get ugly probably

Reply:
I'm more on the side of leaving this as-is if that works for you?

Reply:
Yeah, let's leave it as is